File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int n = nums.size();
        int zero = 0;
        int j = 0;
        int i = 0;
        int ans = INT_MIN;
        while(j<n) {
            
            if(nums[j]==0) zero++;

            if(zero>k) {
                if(nums[i]==0) zero--;
                i++;
            }

            if(zero<=k)
                ans = max(ans,j-i+1);

            j++;
        }

        return ans;
    }
};

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int n = nums.length;
        int i = 0;
        int j = 0;
        int z = 0;
        int ans = 0;
        Queue<Integer> que = new ArrayDeque<>();

        while(j<n) {

            if(nums[j]==0) {
                z++;
                que.add(j);
            }
            
            if(z>k) {
                ans = Math.max(ans,j-i);
                i = Math.max(i,que.poll()+1);
                z--;
            } 

            j++;
        }
        ans = Math.max(ans,j-i);
        return ans;
    }
}